pkg/protostore/protostore.go

```diff
@@ -80,10 +80,10 @@ type ProtoStoreMsg struct {
 	Msg        []byte
 	ParentId   string
 	Revision   int64
-	OrgId      string `gorm:"primaryKey"`
-	InstanceId string `gorm:"primaryKey"`
-	CreatedAt  time.Time
-	UpdatedAt  time.Time
+	OrgId      string    `gorm:"primaryKey"`
+	InstanceId string    `gorm:"primaryKey"`
+	CreatedAt  time.Time `gorm:"autoUpdateTime:false;default:current_timestamp"`
+	UpdatedAt  time.Time `gorm:"autoUpdateTime:false;default:current_timestamp"`
```

Comment:
The result of adding `gorm:"autoUpdateTime:false;default:current_timestamp"` tag to `UpdatedAt` is gonna be that if consumer passes a specific _updated_at_ value, then that value is gonna be persisted in the DB. If consumer passes a zero timestamp as _updated_at_, then Gorm will set current timestamp as _updated_at_. Same goes for _created_at_ column.

If I didn't add this tag, then Gorm would always set current timestamp as _created_at_/_updated_at_, even if consumer passed a non-zero timestamp.

Reply:
@jeyhun Why is time.now() required?
lets not overload the same filed with autoUpdate and external value like transition_time. Would this allow updatedTime to be go back in timeline too?

Reply:
If you are referring to why I am using `time.Now()` in[ `Update()`](https://github.com/vmware-labs/multi-tenant-persistence-for-saas/blob/set-updated-at-timestamp/pkg/protostore/protostore.go#L190) and [`Upsert()`](https://github.com/vmware-labs/multi-tenant-persistence-for-saas/blob/set-updated-at-timestamp/pkg/protostore/protostore.go#L229), suppose this scenario:
**I** Consumer calls `InsertWithMetadata` and provides _updated_at_ timestamp equal to 1999-12-31 23:59:59
**II** Consumer calls `Update` (and is not able to provide _updated_at_ timestamp because that method doesn't accept metadata as argument).
The expectation, in my opinion, should be that the _updated_at_ column should equal to the current time.

If my `Update` method doesn't call `time.Now()` and set its output to `md.UpdatedAt`, then _updated_at_ column is not going to change (because Gorm is going to set a value for it only if the `UpdatedAt` field passed to `DataStore`'s `Update` method equals zero). So, I have no choice but to manually set UpdatedAt field's value on lines #190 & #229 of protostore.go.

My unit test `TestProtoStoreInDbTimestamps` tests the scenario I described on lines #546-552 of protostore_test.go. If you comment out line #190 of protostore.go and run that unit test, it's gonna fail.

Reply:
> lets not overload the same filed with autoUpdate and external value like transition_time

I am actually disabling autoupdate (that's why it says _autoUpdateTime:false_). Only if a record is being created or modified and no value is provided for _created_at_ and _updated_at_ columns, then the values of those columns will be set to current timestamp.

> Would this allow updatedTime to be go back in timeline too?

It will be possible for the caller to set any _updated_at_, including some timestamp in the past, or to modify the record and set _updated_at_ to an earlier timestamp.

Reply:
These fields are specificly used for api creation and update times, lets not overload them for other usecases.

Reply:
These fields are specificly used for api creation and update times, lets not overload them for other usecases.